rs/xnet/payload_builder/src/certified_slice_pool.rs

```diff
@@ -377,7 +377,18 @@ impl Payload {
         message_limit: Option<usize>,
         byte_limit: Option<usize>,
     ) -> CertifiedSliceResult<(Option<Self>, Option<Self>)> {
-        let message_limit = message_limit.unwrap_or(usize::MAX);
+        // `messages_end` may not exceed a certain `max_messages_index` in order to cap the number
+        // of signals in the reverse stream. Calculate the maximum number of messages we can include
+        // in the slice such that this is respected.
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_messages_index = max_messages_index(self.header.begin()).get() as usize;
+            max_messages_index.saturating_sub(messages_begin)
+        };
+        let message_limit = message_limit.map_or(max_message_limit, |message_limit| {
+            message_limit.min(max_message_limit)
+        });
```

Comment:
This seems a bit more convoluted than necessary, how about something like:
```suggestion
        let max_message_limit = {
            let nr_existing_signals = self.messages_begin.map_or(0, |mb| mb - self.header.begin()) as usize;
            MAX_STREAM_MESSAGES - nr_existing_signals
        };
        let message_limit = message_limit.unwrap_or(max_message_limit).min(max_message_limit);
```
(probably doesn't compile, but you get the idea)

Reply:
I like my version better for a couple of reasons:
- It has only one subtraction in it to worry about.
- It doesn't use the constant `MAX_STREAM_MESSAGES` allowing me to encapsulate this in the function `max_message_index()`.
- It follows a logic of 3 places on the stream index axis I find easier to understand.
I had the same mental picture that your suggestion is based on initially, but the picture of a stream index axis seems cleaner to me since it also fits better with the check in `validate_slice()`.

I mean something like this.
```
--------|------------------|-------------------------|---------------------> stream index
 stream_begin        messages_begin             max_message_index
```
So we have a `max_message_index` up to which we can have signals. After gc'ing up to `stream_begin`, we have signals from there to `messages_begin`. Therefore we can produce `max_message_index - messages_begin` more signals. The `saturating_sub()` is in there for the case where you decrease the number of signals in streams in which case we could have `max_message_index < messages_begin`.

Actually why don't I just put this into the comment above the `max_message_index()` function?

Reply:
Personally I didn't find the `max_message_index` concept particularly helpful, when looking at the code I had to unfold the `max_message_index`definition to figure out what's going on and how it's linked to the motivation of "we don't want too many signals", both here and in `validate_slice`. So I'd find it easier to just get rid of it completely. But you'll have to stare at this code more than I will so I won't push for it, feel free to resolve.